Integration of Hugging Face Hub with Galaxy (#3515)
galaxyproject/galaxy-hub: content/news/2026-01-07-hf-integration/index.md

date: "2026-01-07"
authors: Anup Kumar
tease: "Galaxy users can now browse the Hugging Face Hub as a repository source, import models straight into their histories, and feed them into tools. A step-by-step example shows how to pull models from the Hugging Face Hub into Galaxy and then using the existing DocLayout-YOLO tool for document layout segmentation."
subsites: [global,eu,us]
main_subsite: freiburg
tags: [tools]
---

# Integration of Hugging Face Hub with Galaxy

A common friction point in ML-powered scientific analysis in Galaxy is **getting the appropriate model into the right tool**: downloading weights locally, uploading again, and repeating for every history or workflow run.

With Galaxy’s new [**Hugging Face Hub**](https://huggingface.co/) integration, you can **browse Hugging Face model repositories from within Galaxy** file uploader, import a model artifact directly into your history, and immediately use it as a tool input. This feature is implemented as a Galaxy file source plugin built on the `fsspec` and `huggingface_hub` libraries. 

<div align="center">
    <img src="hf_gpt_oss.png" alt="HF GPT OSS 120B" width="600"/>
</div>

> A wide range of models across many scientific disciplines include protein language models ([Rostlab/prot_bert](https://huggingface.co/Rostlab/prot_bert)), biomedical and clinical NLP ([medicalai/ClinicalBERT](https://huggingface.co/medicalai/ClinicalBERT)), chemistry language models ([ibm-research/MoLFormer-XL-both-10pct](https://huggingface.co/ibm-research/MoLFormer-XL-both-10pct)), weather monitoring models ([nvidia/fourcastnet3](https://huggingface.co/nvidia/fourcastnet3) and [ibm-nasa-geospatial/Prithvi-EO-1.0-100M](https://huggingface.co/ibm-nasa-geospatial/Prithvi-EO-1.0-100M)),  astronomy ([AstroMLab/AstroSage-8B](https://huggingface.co/AstroMLab/AstroSage-8B)), large language models ([Llama-4: meta-llama/Llama-4-Scout-17B-16E-Instruct](https://huggingface.co/meta-llama/Llama-4-Scout-17B-16E-Instruct)) and many more. All of these models can be swiftly pulled from Hugging Face Hub into Galaxy history leading to faster ML-based analyses.

Reviewer: ```suggestion
A wide range of models across many scientific disciplines include protein language models ([Rostlab/prot_bert](https://huggingface.co/Rostlab/prot_bert)), biomedical and clinical NLP ([medicalai/ClinicalBERT](https://huggingface.co/medicalai/ClinicalBERT)), chemistry language models ([ibm-research/MoLFormer-XL-both-10pct](https://huggingface.co/ibm-research/MoLFormer-XL-both-10pct)), weather monitoring models ([nvidia/fourcastnet3](https://huggingface.co/nvidia/fourcastnet3) and [ibm-nasa-geospatial/Prithvi-EO-1.0-100M](https://huggingface.co/ibm-nasa-geospatial/Prithvi-EO-1.0-100M)),  astronomy ([AstroMLab/AstroSage-8B](https://huggingface.co/AstroMLab/AstroSage-8B)), large language models ([Llama-4: meta-llama/Llama-4-Scout-17B-16E-Instruct](https://huggingface.co/meta-llama/Llama-4-Scout-17B-16E-Instruct)) and many more. All of these models can be swiftly pulled from Hugging Face Hub into Galaxy history, leading to faster ML-based analyses.
```

Now you can browse/search Hugging Face Hub models directly from Galaxy. Here we search for the model using *doclayout* as the search string and select the pretrained DocLayout-YOLO repository (juliozhao/DocLayout-YOLO-DocLayNet-Docsynth300K_pretrained).

<div align="center">
    <img src="3_doclayout_hf.png" alt="Browse Hugging Face repos from Galaxy and select the DocLayout-YOLO repository" width="600"/>
</div>

---

### 4) Select the model weights file

Inside the repository, pick the weights file (here a `.pt` file).

<div align="center">
    <img src="4_select_model.png" alt="Select the YOLO weights file from the repository" width="600"/>
</div>

---

### 5) Import the selected file into your history

After adding the file to the upload queue, click **Start** to import it into your Galaxy history.

<div align="center">
    <img src="5_import_model.png" alt="File added to upload queue; click Start to import" width="600"/>
</div>

At this point, your YOLO weights are now a normal Galaxy dataset: usable as a tool/workflow input and shareable.

---

## Run inference in Galaxy: DocLayout-YOLO segmentation

### 6) Configure the DocLayout-YOLO tool

> Open the **DocLayout-YOLO** tool form and set:

Reviewer: ```suggestion
Open the [**DocLayout-YOLO**](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/bgruening/doclayoutyolo/doclayoutyolo/0.0.4.1+galaxy0) tool form and set:
```

Reviewer: How do I get to this short link version? Then I will do my best to remember for the next post :)

- **Yolo model** → the `.pt` file imported from Hugging Face
- **Input image** → the document page image to be segmented

<div align="center">
    <img src="6_doclayout_tool.png" alt="DocLayout-YOLO tool form using the imported model weights" width="600"/>
</div>

Then click **Run Tool**.

---

### 7) View the output: segmented page with detected layout regions

The tool produces a segmented image with detected regions (e.g., headings, paragraphs, page headers) drawn as bounding boxes.

<div align="center">
    <img src="7_segmented_image.png" alt="Segmented output image produced by DocLayout-YOLO" width="600"/>
</div>

> 

Reviewer 2: You can use [🔧 Convert coordinates to label map](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/points2labelimage/ip_points_to_label) to convert GeoJSON to a Label map, followed by [🔧 Crop image](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/crop_image/ip_crop_image) to extract the image patches corresponding to the labeled regions (the bounding boxes of those, which are identical to the annotated image regions, if they are rectangular).

> 

Reviewer 2: @anuprulez I see! Thanks for pointing this out. I will have a look at the Crop image tool, this should be easy to fix. In the mean time, you can use one of two workarounds.

If you need RGB images, you can use the [🔧 Split image along axes](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/split_image/ip_split_image/2.3.5+galaxy0) tool to split the RGB image into a list of single-channel images, apply the [🔧 Crop image](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/crop_image/ip_crop_image) tool, and then re-combine the RGB stacks using the [🔧 Concatenate images or channels](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/concat_channels/ip_concat_channels/0.5) tool.

If you're fine with grayscale images, again, use the [🔧 Split image along axes](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/split_image/ip_split_image/2.3.5+galaxy0) tool to split the RGB image into a list of single-channel images, and then just use one of the single-channel images for the [🔧 Crop image](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/crop_image/ip_crop_image) tool. Or, use the [Process images using arithmetic expressions](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/image_math/image_math/1.26.4+galaxy2) tool to combine the single-channel images into a single grayscale image by averaging over the channel images.

> 

Author: Workflow: https://usegalaxy.eu/u/kumara/w/ocr-with-doclayout-hugging-face-and-llm-hub-1

Invocation: https://usegalaxy.eu/workflows/invocations/6fc32b7a39dc5b6e

History: https://usegalaxy.eu/u/kumara/h/ocr-doclayout-hugging-face-llm-hub

I will update the post with this workflow instead of showing each step one-by-one

---

## Significance for reproducible Galaxy ML-based workflows

Pulling models from Hugging Face from *inside Galaxy* makes ML workflows much easier to operationalize:

- **Broad model coverage**: A wide range of models such as large language models (LLMs), biology and imaging foundation models and many more are available for downstream analyses.
- **No manual download/upload loop** (especially cumbersome for large weights and models)
- **Tool inputs become explicit**: the exact model file used is captured in the history and workflow provenance.
- **Reusable across analyses**: share histories/workflows with the model artifact included.

> - **Standardisation of ML models**: usage of widely used Hugging Face Hub as model repository provides standard and safer model loaders and structured model metadata.

Reviewer: ```suggestion
- **Standardization of ML models**: usage of widely used Hugging Face Hub as model repository provides standard and safer model loaders and structured model metadata.
```

for unification, as you used "organize" instead of organise earlier.

<div align="center">
    <img src="yolo_workflow.png" alt="Segmented output image produced by DocLayout-YOLO" width="400"/>
</div>

### Configure the DocLayout-YOLO tool in the workfow

The [**DocLayout-YOLO**](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/bgruening/doclayoutyolo/doclayoutyolo/0.0.4.1+galaxy0) tool form and set:

- **Yolo model** → the `.pt` file imported from Hugging Face
- **Input image** → the document page image to be segmented

#### The segmented image with detected layout regions

The DocLayout-YOLO tool produces a segmented image with detected regions (e.g., headings, paragraphs, page headers) drawn as bounding boxes.

<div align="center">
    <img src="7_segmented_image.png" alt="Segmented output image produced by DocLayout-YOLO" width="400"/>
</div>

### Configure the LLM Hub tool in the workfow

The LLM Hub tool utilises multi-modal LLMs with adavanced OCR capabilities to extract text from the images having bounded regions. The tool has the following settings.

<div align="center">
    <img src="llm_hub_step.png" alt="Segmented output image produced by DocLayout-YOLO" width="400"/>
</div>

### Workflow output

> The following image shows the output of the text segmentation and detection output produced by the workflow in the markdown format. Additionally, the output enlists the **thinking process** of the associated imaging LLM before producing the text from the bounding regions. The [workflow invocation](https://usegalaxy.eu/workflows/invocations/6fc32b7a39dc5b6e) provides more details.

Reviewer: ```suggestion
The following image shows the output of the text segmentation and detection produced by the workflow. Depending on the input image, we get several separate images. Those can now be used with other tools, like Tesseract or LLM Hub, for example.  
```
